Re-work the class hierarchy of cursors and transactions

horazont opened the pull request:
MDBRWTransaction now inherits from MDBROTransaction. Both
transaction types will free their cursors before aborting or
committing. In addition, transactions are now virtual classes.

The reason for this is that RW and RO transactions are handled
very differently in LMDB. Nevertheless, it is very useful to be
able to write read-only code in a way which also can use an RW
transaction. This saves code duplication or unnecessary templates.

Since the only methods which are reqiured to be virtual on
transactions for this to work are the destructor and the
abort/commit methods, the overhead should be neglegible.

This also comes in very handy when going forward to implement
nested transactions, because it is not possible to nest RO
transactions below RW transactions, exacerbating the issue
described above.

Fixes #7 en passant.

horazont committed:
Re-work the class hierarchy of cursors and transactions

MDBRWTransaction now inherits from MDBROTransaction. Both
transaction types will free their cursors before aborting or
committing. In addition, transactions are now virtual classes.

The reason for this is that RW and RO transactions are handled
very differently in LMDB. Nevertheless, it is very useful to be
able to write read-only code in a way which also can use an RW
transaction. This saves code duplication or unnecessary templates.

Since the only methods which are reqiured to be virtual on
transactions for this to work are the destructor and the
abort/commit methods, the overhead should be neglegible.

This also comes in very handy when going forward to implement
nested transactions, because it is not possible to nest RO
transactions below RW transactions, exacerbating the issue
described above.

Fixes #7 en passant.

Habbie commented:
Thank you - based on the description, this should go a long way to fixing https://github.com/PowerDNS/pdns/issues/8134 which currently makes the PowerDNS LMDB backend somewhat useless. Is this PR in a shape where it would make sense for me to test it in PowerDNS? Thanks!

horazont commented:
Given that the tests came back green (both the Catch ones and the examples), I think that would be great.

That way, any issues and quirks caused by the remodel can be caught quickly and while/before I build nesting on top of it.

Thanks for looking at it.

horazont commented:
I just thought of an issue though, which will need a more invasive change. Consider this code:

```cpp
MDBRWTransaction txn = env.getRWTransaction();

// Invalid: We explicitly break this move because it would be unsafe:
// MDBROTransaction ro_txn(std::move(txn));

// Valid, RW inherits from RO now, so we can bind an RO reference to an RW 
// transaction.
MDBROTransaction &ro_txn = txn;

// Dangerous!!
MDBROTransaction ro_txn2(std::move(ro_txn));
```

The last move there breaks the semantics of the RW transaction which is bound to the reference ro_txn. It looses its RW cursors, which remain partly inside the ``txn`` instance. All kinds of weird and bad things can happen here. For instance, the ``ro_txn2`` would go out of scope before the ``txn``, calling the destructor MDBROTransaction destructor (which defaults to commit instead of abort!) and only freeing parts of the cursors. Only then the MDBRWTransaction destructor is called, which will free the cursors which belong to the RW transaction which has already been committed.

This effect is called ["Object Slicing"](https://en.wikipedia.org/wiki/Object_slicing); I am surprised to find a wikipedia article about it.

Here’s how I would solve this: Disallow moving of transactions altogether. Instead, the API returns std::unique_ptr-wrapped transactions. This means that we need a dynamic allocation per transaction, however, this would allow us to avoid the dynamic allocation for the std::vector std::unique_ptr wrappers we currently use for the cursor lists.

This of course breaks the API quite a bit, because now the transactions are wrapped in unique_ptr.

We can mitigate the breakage a bit by renaming the classes and providing aliases:

```cpp
using MDBROTransaction = std::unique_ptr<MDBROTransactionImpl>;
using MDBRWTransaction = std::unique_ptr<MDBRWTransactionImpl>;
```

Or similar.

horazont committed:
Hide MDB*Transaction behind a unique_ptr front

This is to prevent the issue with Object Slicing. With the previous
solution (where MDB*Transaction are normal objects), consider the
following code:

    MDBRWTransaction txn = env.getRWTransaction();

    //! Invalid: We explicitly break this move because it would be
    //! unsafe:
    // MDBROTransaction ro_txn(std::move(txn));

    //! Valid, RW inherits from RO now, so we can bind an RO
    //! reference to an RW transaction.
    MDBROTransaction &ro_txn = txn;

    //! Dangerous!!
    MDBROTransaction ro_txn2(std::move(ro_txn));

The last move there breaks the semantics of the RW transaction which
is bound to the reference ro_txn. It looses its RW cursors, which
remain partly inside the txn instance. All kinds of weird and bad
things can happen here. For instance, the ro_txn2 would go out of
scope before the txn, calling the destructor MDBROTransaction
destructor (which defaults to commit instead of abort!) and only
freeing parts of the cursors. Only then the MDBRWTransaction
destructor is called, which will free the cursors which belong to
the RW transaction which has already been committed.

The only safe way to prevent Object Slicing in this scenario I
could come up with is to disallow moves of the objects altogether
and instead use unique_ptr as front for them. This also removes
an additional dynamic allocation per RW transaction (for the
cursor vector), since the address of that vector is now constant
over the lifetime of the transaction without indirection.
